randn op support global tensor

grybd opened the pull request:
此pr目的：
- [x] randn(normal) op 支持 2D SBP
- [x] randn op 支持 global tensor 、添加单测

grybd commented:
normal op 支持 global tensor 的PR已经ready了麻烦帮忙review下~ @strint @clackhan @wyg1997

strint commented:
需要完善一下测试的数值对比

grybd commented:
@wyg1997 @clackhan 麻烦帮忙review以下哈

grybd commented:
_____________________ TestRandnConsistent.test_randn_graph _____________________
[gw10] linux -- Python 3.7.10 /opt/conda/bin/python3

test_case = <test_consistent_randn.TestRandnConsistent testMethod=test_randn_graph>

    @globaltest
    def test_randn_graph(test_case):
        arg_dict = OrderedDict()
        arg_dict["shape"] = [(8,), (8, 8,), (8, 8, 8)]
        arg_dict["placement"] = [
            # 1d
            flow.placement("cpu", ranks=[0, 1]),
>           flow.placement("cuda", ranks=[0, 1]),
            # 2d
            flow.placement("cpu", ranks=[[0, 1],]),
            flow.placement("cuda", ranks=[[0, 1],]),
        ]
E       oneflow._oneflow_internal.exception.RuntimeError: placement type cuda not supported.

wyg1997 commented on python/oneflow/test/modules/test_consistent_randn.py:
```diff
@@ -0,0 +1,122 @@
+    test_case.assertEqual(x.placement, placement)
+
+
+class TestRandnConsistent(flow.unittest.TestCase):
```
加一个 skip_if，在 CPU_ONLY 的情况下跳过。

grybd replied:
好的~

grybd commented:
Error: (GRAPH:ConsistentRandnGraph_0:ConsistentRandnGraph) building graph got error: <class 'oneflow._oneflow_internal.exception.Exception'> 
  File "/home/ci-user/runners/release/_work/oneflow/oneflow/oneflow/core/functional/impl/random_functor.cpp", line 185, in operator()
    CheckDeviceIdsIsValid(placement)
  File "/home/ci-user/runners/release/_work/oneflow/oneflow/oneflow/core/job/parallel_desc.cpp", line 499, in RawCheckDeviceIdsIsValid
    placement->CheckDeviceIdsIsValid()
  File "/home/ci-user/runners/release/_work/oneflow/oneflow/oneflow/core/job/parallel_desc.cpp", line 323, in CheckDeviceIdsIsValid
    Check failed: (machine_id) < (GlobalProcessCtx::WorldSize()) (1 vs 1) Placment is invalid because rank must be less than world size!

@strint @wyg1997 这是什么原因呢

grybd commented:
=================================== FAILURES ===================================
__________________ TestConvertDependency.test_infos_of_nodes ___________________
[gw13] linux -- Python 3.7.11 /opt/conda/bin/python3

test_case = <test_tvm_frontend_dependency_on_graph.TestConvertDependency testMethod=test_infos_of_nodes>

    def test_infos_of_nodes(test_case):
        alexnet_module = alexnet()
        alexnet_graph = Graph(alexnet_module)
        if not alexnet_graph._is_compiled:
            alexnet_graph._compile(flow.rand(1, 3, 224, 224))
        graph_str = repr(alexnet_graph)
        if not alexnet_graph._is_compiled:
            alexnet_graph._compile(flow.rand(shape_input))
    
        size_where = 2
        if "cuda" in graph_str:
            size_where = 3
    
        p_size = re.compile(r"size=\(.*?\)", re.S)
        p_type = re.compile(r"dtype=.*?\)", re.S)
        types = ["INPUT", "PARAMETER", "BUFFER", "OUTPUT"]
        num_nodes = {}
    
        for t in types:
            data = re.finditer(t + ":.*", graph_str)
            cnt = 0
            for i in data:
                cnt += 1
                attrs = i.group().split(":")
                size_strs = re.findall(p_size, attrs[size_where])
                type_strs = re.findall(p_type, attrs[size_where])
                test_case.assertEqual(size_strs != [], True)
                test_case.assertEqual(type_strs != [], True)
    
                size_attr = size_strs[0].replace("size=", "")
                type_attr = type_strs[0].replace("dtype=", "").replace(")", "")
                if size_attr[-2] == ",":
                    size_attr = size_attr.replace(",", "")
                if type_attr[-1] == ",":
                    type_attr = type_attr.replace(",", "")
                    test_case.assertEqual(type_attr, "oneflow.float32")
    
                data_size = tuple(map(int, size_attr[1:-1].split(", ")))
                if cnt == 1 and t == "PARAMETER":
                    test_case.assertEqual(data_size, (64, 3, 11, 11))
                elif cnt == 15 and t == "PARAMETER":
                    test_case.assertEqual(data_size, (1000, 4096))
            num_nodes[t] = cnt
    
        test_case.assertEqual(num_nodes["INPUT"] != 0, True)
        test_case.assertEqual(num_nodes["BUFFER"], 0)
        test_case.assertEqual(num_nodes["PARAMETER"], 16)
        test_case.assertEqual(num_nodes["OUTPUT"] != 0, True)
    
        # get graph proto, if you don't _compile the graph, the _graph_proto will be None
        graph_input = re.search(r"INPUT:.*", graph_str).group().split(":")
        shape_input = tuple(
            map(
                int,
                re.findall(p_size, graph_input[size_where])[0]
                .replace("size=", "")[1:-1]
                .split(", "),
            )
        )
        graph_proto = alexnet_graph._graph_proto
    
        nodes = {}
        for op in graph_proto.net.op:
            nodes[op.name] = op
    
        op_names = []
        op_attrs = []
        for node_name in nodes:
            node = nodes[node_name]
            if is_user_op(node):
                op_name = node.user_conf.op_type_name
                op_attr = parse_attr(node.user_conf.attr)
                op_names.append(op_name)
                op_attrs.append(op_attr)
    
        test_case.assertEqual(op_names[0], "conv2d")
        test_case.assertEqual(op_names[1], "bias_add")
        test_case.assertEqual(op_names[2], "relu")
    
        kernel_size = op_attrs[0].get("kernel_size", None)
        strides = op_attrs[0].get("strides", None)
        padding_before = op_attrs[0].get("padding_before", None)
        test_case.assertEqual(kernel_size, (11, 11))
        test_case.assertEqual(strides, (4, 4))
        test_case.assertEqual(padding_before, (2, 2))
    
        node_input_list = []
        node_output_list = []
        for node_name in nodes:
            node = nodes[node_name]
            if is_user_op(node) and node.user_conf.op_type_name == "conv2d":
                for input_name in node.user_conf.input:
                    node_input_paths = getattr(node.user_conf.input[input_name], "s")
                    for i in node_input_paths:
                        node_input = i.split("/")[0]
                        print(node_input)
                        node_input_list.append(node_input)
                for output_name in node.user_conf.output:
                    node_output_paths = getattr(node.user_conf.output[output_name], "s")
                    for node_output_path in node_output_paths:
                        node_output_name = node_output_path.split("/")[0]
                        print(node_output_name)
                        node_output_list.append(node_output_name)
    
>       test_case.assertEqual("_Graph_1_input.0.0_2" in node_input_list, True)
E       AssertionError: False != True

python/oneflow/test/graph/test_tvm_frontend_dependency_on_graph.py:[202](https://github.com/Oneflow-Inc/oneflow/runs/6078026948?check_suite_focus=true#step:35:202): AssertionError

grybd commented:
_____________________ TestRandnConsistent.test_randn_graph _____________________
[gw0] linux -- Python 3.7.11 /opt/conda/bin/python3

test_case = <test_consistent_randn.TestRandnConsistent testMethod=test_randn_graph>

    @unittest.skipIf(os.getenv("ONEFLOW_TEST_CPU_ONLY"), "only test cpu cases")
    @globaltest
    def test_randn_graph(test_case):
        arg_dict = OrderedDict()
        arg_dict["shape"] = [(8,), (8, 8,), (8, 8, 8)]
        arg_dict["placement"] = [
            # 1d
            flow.placement("cpu", ranks=[0, 1]),
            flow.placement("cuda", ranks=[0, 1]),
            # 2d
            flow.placement("cpu", ranks=[[0, 1],]),
            flow.placement("cuda", ranks=[[0, 1],]),
        ]
        for args in GenArgDict(arg_dict):
            shape = args["shape"]
            placement = args["placement"]
            for sbp in all_sbp(placement, max_dim=len(shape), except_partial_sum=True):
>               _test_graph_randn(test_case, shape, placement, sbp)

python/oneflow/test/modules/test_consistent_randn.py:119: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
python/oneflow/test/modules/test_consistent_randn.py:80: in _test_graph_randn
    x = model()
/opt/conda/lib/python3.7/site-packages/oneflow/nn/graph/graph.py:214: in __call__
    self._compile(*args, **kwargs)
/opt/conda/lib/python3.7/site-packages/oneflow/nn/graph/graph.py:660: in _compile
    eager_outputs = self.__build_graph(*args, **kwargs)
/opt/conda/lib/python3.7/site-packages/oneflow/nn/graph/graph.py:751: in __build_graph
    outputs = self.build(*lazy_args, **lazy_kwargs)
python/oneflow/test/modules/test_consistent_randn.py:76: in build
    x = flow.randn(*shape, placement=placement, sbp=sbp)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

out = None, generator = None, dtype = None, layout = None

    def randn_op(
        *size,
        out=None,
        generator=None,
        dtype: Optional[flow.dtype] = None,
        layout=None,
        device: Union[flow.device, str, None] = None,
        placement: flow.placement = None,
        sbp: flow._oneflow_internal.sbp.sbp = None,
        requires_grad: bool = False
    ):
        """
        Returns a tensor filled with random numbers from a normal distribution with mean 0 and variance 1 (also called the standard normal distribution).
    
        The shape of the tensor is defined by the variable argument ``size``.
    
        Args:
            size (int... or oneflow.Size): Defining the shape of the output tensor.
              Can be a variable number of arguments or a collection like a list or tuple or oneflow.Size.
            out (optional): The output tensor.
            dtype (flow.dtype, optional): The desired data type of returned tensor. Default: ``flow.float32``.
            layout (optional): The desired layout of returned Tensor.
            generator (flow.Generator, optional): a pseudorandom number generator for sampling
            device (flow.device, optional): The desired device of returned local tensor. If None, uses the
              current device.
            placement (flow.placement, optional): The desired device of returned global tensor. If None, will
              construct local tensor.
            sbp (flow.sbp, optional): The desired sbp of returned global tensor. It must be equal with the
              numbers of placement.
            requires_grad (bool, optional): If autograd should record operations on the returned tensor. Default: False.
    
        For example:
    
        .. code-block:: python
    
            >>> import oneflow as flow
            >>> x = flow.randn(3,3) # construct local tensor
            >>> x.shape
            oneflow.Size([3, 3])
            >>> x.is_global
            False
            >>> placement = flow.placement("cpu", ranks=[0])
            >>> sbp = flow.sbp.broadcast
            >>> x = flow.randn(3,3,placement=placement,sbp=sbp) # construct global tensor
            >>> x.is_global
            True
    
        """
        size = _handle_size_arg(size)
        assert out is None, "out not supported yet"
        assert layout is None, "layout not supported yet"
        if placement is not None:
            return flow._C.randn(
                size=size,
                placement=placement,
                sbp=sbp,
                dtype=dtype,
                generator=generator,
>               requires_grad=requires_grad,
            )
E           oneflow._oneflow_internal.exception.Exception: 
E             File "/home/ci-user/runners/release/_work/oneflow/oneflow/oneflow/core/functional/impl/random_functor.cpp", line [185](https://github.com/Oneflow-Inc/oneflow/runs/6133527331?check_suite_focus=true#step:36:185), in operator()
E               CheckDeviceIdsIsValid(placement)
E             File "/home/ci-user/runners/release/_work/oneflow/oneflow/oneflow/core/job/parallel_desc.cpp", line 499, in RawCheckDeviceIdsIsValid
E               placement->CheckDeviceIdsIsValid()
E             File "/home/ci-user/runners/release/_work/oneflow/oneflow/oneflow/core/job/parallel_desc.cpp", line 323, in CheckDeviceIdsIsValid
E               Check failed: (machine_id) < (GlobalProcessCtx::WorldSize()) (1 vs 1) Placment is invalid because rank must be less than world size!

/opt/conda/lib/python3.7/site-packages/oneflow/nn/modules/random_ops.py:[208](https://github.com/Oneflow-Inc/oneflow/runs/6133527331?check_suite_focus=true#step:36:208): Exception
----------------------------- Captured stdout call -----------------------------
Error: (GRAPH:ConsistentRandnGraph_0:ConsistentRandnGraph) building graph got error: <class 'oneflow._oneflow_internal.exception.Exception'> 
  File "/home/ci-user/runners/release/_work/oneflow/oneflow/oneflow/core/functional/impl/random_functor.cpp", line 185, in operator()
    CheckDeviceIdsIsValid(placement)
  File "/home/ci-user/runners/release/_work/oneflow/oneflow/oneflow/core/job/parallel_desc.cpp", line 499, in RawCheckDeviceIdsIsValid
    placement->CheckDeviceIdsIsValid()
  File "/home/ci-user/runners/release/_work/oneflow/oneflow/oneflow/core/job/parallel_desc.cpp", line 323, in CheckDeviceIdsIsValid
    Check failed: (machine_id) < (GlobalProcessCtx::WorldSize()) (1 vs 1) Placment is invalid because rank must be less than world size!
